## Group Assignment #2: Census Data Exploration

###### For this assignment I will be exploring the census data of Los Angles, and evlauating its racial profile. This data is important because our goal is to understand how socioeconomic profiles in LA may impact their vunerbility to the extremities of heat. My goal is to seek the percentage of minorities in the census blocks, and compare minority population to the entirety of the city. 

###### Author of the assignment: Pa-Shun Hawkins 

### Importing Library and Data

###### In this first step, I will be importing pandas because it is a library, and this will allow me to work with the data. The out come of this is the pandas opening. 

In [2]:
#impoting the library pandas
import pandas as pd

In [3]:
# Here I am loading the data file into python and renaming it df
# I am also typing out "df" which will print a small sample of the data
#for me to see
df=pd.read_csv('Data/la_demographincs.csv')
df

Geo_FIPS              Geo_NAME  \
0     6037101110  Census Tract 1011.10   
1     6037101122  Census Tract 1011.22   
2     6037101220  Census Tract 1012.20   
3     6037101221  Census Tract 1012.21   
4     6037101222  Census Tract 1012.22   
...          ...                   ...   
2493  6037980038  Census Tract 9800.38   
2494  6037980039  Census Tract 9800.39   
2495  6037990100     Census Tract 9901   
2496  6037990200     Census Tract 9902   
2497  6037990300     Census Tract 9903   

                                              Geo_QNAME  Geo_NATION  \
0     Census Tract 1011.10, Los Angeles County, Cali...           0   
1     Census Tract 1011.22, Los Angeles County, Cali...           0   
2     Census Tract 1012.20, Los Angeles County, Cali...           0   
3     Census Tract 1012.21, Los Angeles County, Cali...           0   
4     Census Tract 1012.22, Los Angeles County, Cali...           0   
...                                                 ...         ...   
2493  Census Tract 9800.38, Los Angeles County, Cali...           0   
2494  Census Tract 9800.39, Los Angeles County, Cali...           0   
2495  Census Tract 9901, Los Angeles County, California           0   
2496  Census Tract 9902, Los Angeles County, California           0   
2497  Census Tract 9903, Los Angeles County, California           0   

      Geo_STATE  Geo_COUNTY  Geo_CS  Geo_CT  Geo_PLACE  Geo_CSA  ...  \
0             6          37     NaN  101110        NaN      NaN  ...   
1             6          37     NaN  101122        NaN      NaN  ...   
2             6          37     NaN  101220        NaN      NaN  ...   
3             6          37     NaN  101221        NaN      NaN  ...   
4             6          37     NaN  101222        NaN      NaN  ...   
...         ...         ...     ...     ...        ...      ...  ...   
2493          6          37     NaN  980038        NaN      NaN  ...   
2494          6          37     NaN  980039        NaN      NaN  ...   
2495          6          37     NaN  990100        NaN      NaN  ...   
2496          6          37     NaN  990200        NaN      NaN  ...   
2497          6          37     NaN  990300        NaN      NaN  ...   

      SE_T017_020  SE_T017_002  SE_T017_004  SE_T017_006  SE_T017_008  \
0        4.121750     4.050484     6.927711   -12.162162   -50.000000   
1        1.746725     3.140175     7.263028    19.230769    75.000000   
2       -0.564972     8.225729     5.354994    -5.555556   -63.636364   
3        0.291206     0.132567     5.444126   -17.977528   -25.000000   
4        9.859155    17.241379    27.074236   -19.148936    50.000000   
...           ...          ...          ...          ...          ...   
2493   -50.757040   -74.477601  -100.000000  -100.000000  -100.000000   
2494          NaN          NaN          NaN          NaN          NaN   
2495          NaN          NaN          NaN          NaN          NaN   
2496          NaN          NaN          NaN          NaN          NaN   
2497          NaN          NaN          NaN          NaN          NaN   

      SE_T017_010  SE_T017_012  SE_T017_014  SE_T017_016  SE_T017_018  
0      -10.699588        100.0   -53.333333    23.776224     4.305136  
1      -20.808081       -100.0   130.000000     2.586207    -5.739130  
2       29.411765       -100.0   600.000000    -3.809524   -15.015723  
3       -5.747126          0.0   100.000000   -41.666667     0.597780  
4      -20.855615          NaN          NaN    -2.500000    -0.763359  
...           ...          ...          ...          ...          ...  
2493  -100.000000          NaN  -100.000000  1062.515898   597.509539  
2494          NaN          NaN          NaN          NaN          NaN  
2495          NaN          NaN          NaN          NaN          NaN  
2496          NaN          NaN          NaN          NaN          NaN  
2497          NaN          NaN          NaN          NaN          NaN  

[2498 rows x 50 columns]

#### Checking the Size of the Data
###### By looking at the shape of the data we will see the number of rows and columns that this data frame has. For this instance there are 2498 rows and 50 columns, and the head function will show the top of the data set

In [4]:
# This will show the rows and columns
df.shape

(2498, 50)

In [5]:
# looking at a subset of the data. This will portray the first 5 rows of data.
df.head()

Geo_FIPS              Geo_NAME  \
0  6037101110  Census Tract 1011.10   
1  6037101122  Census Tract 1011.22   
2  6037101220  Census Tract 1012.20   
3  6037101221  Census Tract 1012.21   
4  6037101222  Census Tract 1012.22   

                                           Geo_QNAME  Geo_NATION  Geo_STATE  \
0  Census Tract 1011.10, Los Angeles County, Cali...           0          6   
1  Census Tract 1011.22, Los Angeles County, Cali...           0          6   
2  Census Tract 1012.20, Los Angeles County, Cali...           0          6   
3  Census Tract 1012.21, Los Angeles County, Cali...           0          6   
4  Census Tract 1012.22, Los Angeles County, Cali...           0          6   

   Geo_COUNTY  Geo_CS  Geo_CT  Geo_PLACE  Geo_CSA  ...  SE_T017_020  \
0          37     NaN  101110        NaN      NaN  ...     4.121750   
1          37     NaN  101122        NaN      NaN  ...     1.746725   
2          37     NaN  101220        NaN      NaN  ...    -0.564972   
3          37     NaN  101221        NaN      NaN  ...     0.291206   
4          37     NaN  101222        NaN      NaN  ...     9.859155   

   SE_T017_002  SE_T017_004  SE_T017_006  SE_T017_008  SE_T017_010  \
0     4.050484     6.927711   -12.162162   -50.000000   -10.699588   
1     3.140175     7.263028    19.230769    75.000000   -20.808081   
2     8.225729     5.354994    -5.555556   -63.636364    29.411765   
3     0.132567     5.444126   -17.977528   -25.000000    -5.747126   
4    17.241379    27.074236   -19.148936    50.000000   -20.855615   

   SE_T017_012  SE_T017_014  SE_T017_016  SE_T017_018  
0        100.0   -53.333333    23.776224     4.305136  
1       -100.0   130.000000     2.586207    -5.739130  
2       -100.0   600.000000    -3.809524   -15.015723  
3          0.0   100.000000   -41.666667     0.597780  
4          NaN          NaN    -2.500000    -0.763359  

[5 rows x 50 columns]

#### Inspect Data Types
###### Looking at the different variables of the data, and the type of data they are. 

In [6]:
# Using verbose to see the columns and show_counts to see the non null variables

df.info (verbose=True, show_counts= True) 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2498 entries, 0 to 2497
Data columns (total 50 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Geo_FIPS     2498 non-null   int64  
 1   Geo_NAME     2498 non-null   object 
 2   Geo_QNAME    2498 non-null   object 
 3   Geo_NATION   2498 non-null   int64  
 4   Geo_STATE    2498 non-null   int64  
 5   Geo_COUNTY   2498 non-null   int64  
 6   Geo_CS       0 non-null      float64
 7   Geo_CT       2498 non-null   int64  
 8   Geo_PLACE    0 non-null      float64
 9   Geo_CSA      0 non-null      float64
 10  SE_T004_019  2498 non-null   int64  
 11  SE_T004_001  2498 non-null   int64  
 12  SE_T004_003  2498 non-null   int64  
 13  SE_T004_005  2498 non-null   int64  
 14  SE_T004_007  2498 non-null   int64  
 15  SE_T004_009  2498 non-null   int64  
 16  SE_T004_011  2498 non-null   int64  
 17  SE_T004_013  2498 non-null   int64  
 18  SE_T004_015  2498 non-null   int64  
 19  SE_T00

#### Bringing Back the Leading Zeros for the Unique Identifier
###### First we will use the code below to look at the FIPS codes, and verifiy that the leading zeros are missing. The outcome is the first five rows of the Geo_FIPS column, and the zeros are missing.

In [7]:
df.Geo_FIPS.head()

0    6037101110
1    6037101122
2    6037101220
3    6037101221
4    6037101222
Name: Geo_FIPS, dtype: int64

###### Next, I am going to change the dtype of the unique indentifiess from interger to strings. The outcome from the next code will be the dtype of these intergers changing. Also, because it is a string now, the leading zero will be back.

In [8]:
# I am reading in the file, calling the columns and chaning their dtype
df = pd.read_csv(
    'Data/la_demographincs.csv',
    dtype=
    {
        'Geo_FIPS':str,
        'Geo_STATE':str,
        'Geo_COUNTY': str
    }
)

###### The next code will check to see if the intergers were changed to string. This will also reveal if the leading zeros are back.

In [9]:
# showing top of data frame
df.head()

Geo_FIPS              Geo_NAME  \
0  06037101110  Census Tract 1011.10   
1  06037101122  Census Tract 1011.22   
2  06037101220  Census Tract 1012.20   
3  06037101221  Census Tract 1012.21   
4  06037101222  Census Tract 1012.22   

                                           Geo_QNAME  Geo_NATION Geo_STATE  \
0  Census Tract 1011.10, Los Angeles County, Cali...           0        06   
1  Census Tract 1011.22, Los Angeles County, Cali...           0        06   
2  Census Tract 1012.20, Los Angeles County, Cali...           0        06   
3  Census Tract 1012.21, Los Angeles County, Cali...           0        06   
4  Census Tract 1012.22, Los Angeles County, Cali...           0        06   

  Geo_COUNTY  Geo_CS  Geo_CT  Geo_PLACE  Geo_CSA  ...  SE_T017_020  \
0        037     NaN  101110        NaN      NaN  ...     4.121750   
1        037     NaN  101122        NaN      NaN  ...     1.746725   
2        037     NaN  101220        NaN      NaN  ...    -0.564972   
3        037     NaN  101221        NaN      NaN  ...     0.291206   
4        037     NaN  101222        NaN      NaN  ...     9.859155   

   SE_T017_002  SE_T017_004  SE_T017_006  SE_T017_008  SE_T017_010  \
0     4.050484     6.927711   -12.162162   -50.000000   -10.699588   
1     3.140175     7.263028    19.230769    75.000000   -20.808081   
2     8.225729     5.354994    -5.555556   -63.636364    29.411765   
3     0.132567     5.444126   -17.977528   -25.000000    -5.747126   
4    17.241379    27.074236   -19.148936    50.000000   -20.855615   

   SE_T017_012  SE_T017_014  SE_T017_016  SE_T017_018  
0        100.0   -53.333333    23.776224     4.305136  
1       -100.0   130.000000     2.586207    -5.739130  
2       -100.0   600.000000    -3.809524   -15.015723  
3          0.0   100.000000   -41.666667     0.597780  
4          NaN          NaN    -2.500000    -0.763359  

[5 rows x 50 columns]

###### Checking again to make sure that the dtypes have been changed. The outome will be the list of all the variables/columns and their dtype. The hope is to see that the dtype of the identifiers has been changed to interger.

In [10]:
# look at the data types
df.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2498 entries, 0 to 2497
Data columns (total 50 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Geo_FIPS     2498 non-null   object 
 1   Geo_NAME     2498 non-null   object 
 2   Geo_QNAME    2498 non-null   object 
 3   Geo_NATION   2498 non-null   int64  
 4   Geo_STATE    2498 non-null   object 
 5   Geo_COUNTY   2498 non-null   object 
 6   Geo_CS       0 non-null      float64
 7   Geo_CT       2498 non-null   int64  
 8   Geo_PLACE    0 non-null      float64
 9   Geo_CSA      0 non-null      float64
 10  SE_T004_019  2498 non-null   int64  
 11  SE_T004_001  2498 non-null   int64  
 12  SE_T004_003  2498 non-null   int64  
 13  SE_T004_005  2498 non-null   int64  
 14  SE_T004_007  2498 non-null   int64  
 15  SE_T004_009  2498 non-null   int64  
 16  SE_T004_011  2498 non-null   int64  
 17  SE_T004_013  2498 non-null   int64  
 18  SE_T004_015  2498 non-null   int64  
 19  SE_T00

#### Removing All Null Columns 

In [11]:
# .dropna is going to remove the null parts in the dataset. Axis one is used to get rid of the columns and how=all makes sure to get rid of all the empthy columns.  
df = df.dropna(axis=1,how="all")

###### Using this next function to make sure there are no empthy columns of data in this data frame anymore. The result is just brackets beause I have removed all the null data columns with the previous function

In [12]:
df.columns[df.isna().all()].tolist()

[]

#### Choosing which Columns to keep and Renaming them
###### Here I am going to only keep the data from 2020 because it is the most recent census data

In [14]:
# I am keeping the total population data, and all the demographic data from 2020.
#Next I am going to create a new data set where the refined columns and their data will be
columns_to_keep = ['Geo_FIPS',
                   'SE_T004_019',
                   'SE_T004_001',
                   'SE_T004_003',
                   'SE_T004_005',
                   'SE_T004_007',
                   'SE_T004_009',
                   'SE_T004_011',
                   'SE_T004_013',
                   'SE_T004_015',
                   'SE_T004_017']
# add it to a new dataframe
df2 = df[columns_to_keep]

##### Below, we are renaming the columns based on the metadata

In [16]:
# Renaming columns
df2.columns = ['FIPS',
'TotalPop',
'Non Hispanic',
'Non Hispanic White Alone',
'Non Hispanic Black Alone',
'Non Hispanic American Indian and Alaska Native Alone',
'Non Hispanic Asian Alone',
'Non Hispanic Native Hawaiian and Pacific Islander Alone',
'Non Hispanic Other Alone',
'Non Hispanic Multi Race',
'Hispanic']

###### Checking to see if they have been renamed. 

In [18]:
# Printing the new data frame to ensure that the renaming occured
df2

FIPS  TotalPop  Non Hispanic  Non Hispanic White Alone  \
0     06037101110      4926          3545                      2840   
1     06037101122      3728          3186                      2614   
2     06037101220      3344          2263                      1751   
3     06037101221      3444          2266                      1840   
4     06037101222      2808          1768                      1455   
...           ...       ...           ...                       ...   
2493  06037980038         4             2                         0   
2494  06037980039         0             0                         0   
2495  06037990100         0             0                         0   
2496  06037990200         0             0                         0   
2497  06037990300         0             0                         0   

      Non Hispanic Black Alone  \
0                           65   
1                           31   
2                           34   
3                           73   
4                           76   
...                        ...   
2493                         0   
2494                         0   
2495                         0   
2496                         0   
2497                         0   

      Non Hispanic American Indian and Alaska Native Alone  \
0                                                     7      
1                                                     7      
2                                                     4      
3                                                     3      
4                                                     3      
...                                                 ...      
2493                                                  0      
2494                                                  0      
2495                                                  0      
2496                                                  0      
2497                                                  0      

      Non Hispanic Asian Alone  \
0                          434   
1                          392   
2                          352   
3                          246   
4                          148   
...                        ...   
2493                         0   
2494                         0   
2495                         0   
2496                         0   
2497                         0   

      Non Hispanic Native Hawaiian and Pacific Islander Alone  \
0                                                     8         
1                                                     0         
2                                                     0         
3                                                     3         
4                                                     0         
...                                                 ...         
2493                                                  0         
2494                                                  0         
2495                                                  0         
2496                                                  0         
2497                                                  0         

      Non Hispanic Other Alone  Non Hispanic Multi Race  Hispanic  
0                           14                      177      1381  
1                           23                      119       542  
2                           21                      101      1081  
3                           10                       91      1178  
4                            8                       78      1040  
...                        ...                      ...       ...  
2493                         0                        2         2  
2494                         0                        0         0  
2495                         0                        0         0  
2496                         0                        0         0  
2497                         0                        0         0  


In [28]:
df = pd.DataFrame({'Race by Ethnicity':['Non Hispanic Black Alone', 'Non Hispanic American Indian and Alaska Native Alone', 'Non Hispanic Asian Alone'], 'val':[100, 1000, 10000]})

In [29]:
df

Race by Ethnicity    val
0                           Non Hispanic Black Alone    100
1  Non Hispanic American Indian and Alaska Native...   1000
2                           Non Hispanic Asian Alone  10000